sys/vm/swap_pager.c

```diff
@@ -1079,11 +1097,15 @@ swp_pager_xfer_source(vm_object_t srcobject, vm_object_t dstobject,
 	 * swp_pager_meta_build() can sleep.
 	 */
 	VM_OBJECT_WUNLOCK(srcobject);
-	dstaddr = swp_pager_meta_build(dstobject, pindex, addr);
+	dstaddr = swp_pager_meta_build(dstobject, pindex, sb->d[srcmodpi]);
```

Comment:
Access to `sb->d` is serialized by the src object's lock, so this looks a bit iffy to me. It might be possible for a swap_pager_unswapped() call to concurrently free the block, though some other invariants involving the page busy lock probably prevent that today. Nonetheless I'd at least load the block address before dropping the lock, since that might otherwise turn into a hard-to-find bug someday.

Reply:
If a caller frees the block then surely there's already a bug because we shouldn't be using the now-stale block for the destinations swblk?

Reply:
Yes, I don't think there's a bug there at the moment. I just worry a bit that this could turn into a subtle race in the future.

Reply:
I don't really see how you can feasibly introduce such a race accidentally downstream but not upstream... if you want sb to be concurrently modifiable whilst the swap blocks are being transferred then you'd need to re-check that sb->d[i] isn't SWAPBLK_NONE before writing it to the now-allocated destination swblk, freeing the swblk if it is and all the other entries in the destination's are and returning false like the existing check. Anything that would introduce such a race would be obvious and have to conflict in a merge with this code.

Reply:
> Anything that would introduce such a race would be obvious and have to conflict in a merge with this code.

I don't disagree. My observation is really just that this change unnecessarily breaks the invariant that accesses to swap block nodes are protected by the object lock. If you wanted to avoid that, though, it would also make more sense to load the tag bitmap before dropping the source object lock, and then pass both the bitmap and block address to swp_pager_meta_build(). But, that's more invasive and not necessary today, so there's not much point in making the change I suggested originally.

Reply:
Yeah the ifdef-mess that introduces is why I didn't take that approach, but I certainly can see it being useful. I guess from a TSAN perspective it might be better to just load the address under the lock even though we know it's stable.